external/fv3fit/fv3fit/pytorch/recurrent/generator.py

```diff
@@ -0,0 +1,401 @@
+        self.samples_per_day = samples_per_day
+        self.clock = 0
+
+    def reset(self):
+        self.clock = 0
+
+    def forward(self, x):
+        """
+        Args:
+            inputs: tensor of shape [sample, channels, x, y]
+
+        Returns:
+            outputs: tensor of shape [sample, channels + 2, x, y]
+        """
+        features = torch.zeros(x.shape[0], 2, x.shape[2], x.shape[3], device=DEVICE)
+        features[:, 0, :, :] = np.sin(2 * np.pi * self.clock / self.samples_per_day)
+        features[:, 1, :, :] = np.cos(2 * np.pi * self.clock / self.samples_per_day)
+        self.clock = (self.clock + 1) % self.samples_per_day
+        return torch.cat([x, features], dim=-3)
+
+
+class GeographicFeatures(nn.Module):
+    """
+    Appends (x, y, z) features corresponding to Eulerian position of each
+    gridcell on a unit sphere.
+    """
+
+    def __init__(self, nx: int, ny: int):
+        super().__init__()
+        if nx != ny:
+            raise ValueError("this object requires nx=ny")
+        self.xyz = torch.as_tensor(
+            get_grid_xyz(nx=nx).transpose([0, 3, 1, 2]), device=DEVICE
+        ).float()
+
+    def forward(self, x):
+        """
+        Args:
+            x: tensor of shape [sample, tile, channel, x, y]
+
+        Returns:
+            tensor of shape [sample, tile, channel, x, y]
+        """
+        # the fact that this appends instead of prepends is arbitrary but important,
+        # this is assumed to be the case elsewhere in the code.
+        return torch.concat(
+            [x, torch.stack([self.xyz for _ in range(x.shape[0])], dim=0)], dim=-3
+        )
+
+
+class RecurrentGenerator(nn.Module):
+    def __init__(
+        self,
+        config: RecurrentGeneratorConfig,
+        channels: int,
+        nx: int,
+        ny: int,
+        n_time: int,
+        convolution: ConvolutionFactory = single_tile_convolution,
+        step_type: str = "resnet",
+    ):
+        """
+        Args:
+            config: configuration for the network
+            channels: number of input and output channels
+            nx: number of grid points in x direction
+            ny: number of grid points in y direction
+            n_time: number of timesteps in output timeseries, including input timestep
+            convolution: factory for creating all convolutional layers
+                used by the network
+        """
+        super(RecurrentGenerator, self).__init__()
+        self.channels = channels
+        if config.use_geographic_features:
+            xyz_channels = 3
+        else:
+            xyz_channels = 0
+        if config.samples_per_day is not None:
+            diurnal_channels = 2
+        else:
+            diurnal_channels = 0
+        self.hidden_channels = config.max_filters
+        self.nx = nx
+        self.ny = ny
+        self.ntime = n_time
+        self.n_convolutions = config.n_convolutions
+
+        if step_type == "resnet":
```

Comment:
I expect the conv code will be removed once we test it, it did very poorly in my preliminary test. If I work on this code a lot more and we keep the conv code, I'll use the factory pattern I used for the convolution type to avoid the if-elif-else block. Will leave for now.

Reply:
Sounds good.